README.md
# windup: A fast date-time library

`windup` is a library extends from [udatetime](https://pypi.org/project/udatetime/) 
and offers an faster `datetime` object instantiation, serialization and 
deserialization of date-time strings. `windup` is using Python's `datetime class` 
under the hood and code already using `datetime` should be able to easily switch 
to `windup`. All `datetime` objects created by `windup` are timezone-aware. 
The timezones that `windup` uses are fixed-offset timezones, meaning that they 
don't observe daylight savings time (DST), and thus return a fixed offset from 
UTC all year round.

`windup` extends some optional formatting options, all the options are in 
`windup.fmt` and `windup.sep`. 

* windup.fmt
  * fmt.date: formatting date string, like '`2022-07-17`'
  * fmt.time: formatting time string, like '`12:12:12`'
  * fmt.msec: formatting milisecond string, like '12:12:12.`123`'
  * fmt.usec: formatting microsecond string, like '12:12:12.`2123456`'
  * fmt.tz: formatting timezone string, like '2022-07-17T12:12:12.123`+08:00`'
  * fmt.utc_z: formatting timezone string as `Z` if `UTC`, like '2022-07-17T12:12:12.123`Z`'


* windup.sep
  * sep.T: set the separator '`T`' between date-string and time-string
  * sep.space: set the separator '` `' between date-string and time-string
  * sep.underscore: set the separator '`_`' between date-string and time-string



Just see the examples below.

```python
>>> windup.from_string("2021-07-15T12:12:12.123456+08:00")
datetime.datetime(2021, 7, 15, 12, 12, 12, 123456, tzinfo=+08:00)

>>> dt = windup.from_string("2021-07-15T12:12:12.123456+08:00")
>>> windup.to_string(dt)
"2021-07-15T12:12:12.123456+08:00"

>>> windup.now()
datetime.datetime(2021, 7, 15, 12, 12, 12, 472467, tzinfo=+08:00)

>>> windup.utcnow()
datetime.datetime(2021, 7, 15, 4, 12, 12, 472467, tzinfo=+00:00)

>>> windup.now_to_string(option=windup.fmt.date | windup.fmt.time | windup.fmt.usec)
"2021-07-15T12:12:12.123456"

>>> windup.utcnow_to_string(option=windup.fmt.date | windup.fmt.time, sep=windup.sep.space)
"2021-07-15 12:12:12"

>>> windup.fromtimestamp(time.time())
datetime.datetime(2021, 7, 15, 17, 45, 1, 536586, tzinfo=+08:00)

>>> windup.utcfromtimestamp(time.time())
datetime.datetime(2021, 7, 15, 10, 14, 53, tzinfo=+00:00)
```

## Installation

Currently only **POSIX** compliant systems are supported.
Working on cross-platform support.

```
$ pip install windup
```

You might need to install the header files of your Python installation and
some essential tools to execute the build like a C compiler.

```
$ sudo apt-get install python3-dev build-essential
```

or

```
$ sudo yum install python3-devel gcc
```

## Benchmark

```
$ python scripts/bench_windup.py
Executing benchmarks ...

============ benchmark_parse ============
datetime_strptime 2.7934684499996365
windup_parse 0.20049594300053286
windup is 13.9 times faster

============ benchmark_format ============
datetime_strftime 0.6035372909973375
windup_format 0.31113305999315344
windup is 1.9 times faster

============ benchmark_utcnow ============
datetime_utcnow 0.09793541399994865
windup_utcnow 0.04310180398169905
windup is 2.3 times faster

============ benchmark_now ============
datetime_now 0.14119137199941179
windup_now 0.04639216099985788
windup is 3.0 times faster

============ benchmark_utcnow_to_string ============
datetime_utcnow_to_string 0.6921647540002596
windup_utcnow_to_string 0.29203156699077226
windup is 2.4 times faster

============ benchmark_now_to_string ============
datetime_now_to_string 1.2531362709996756
windup_now_to_string 0.49691468400033045
windup is 2.5 times faster

============ benchmark_fromtimestamp ============
datetime_fromtimestamp 0.09680619402206503
windup_fromtimestamp 0.0611271969974041
windup is 1.6 times faster

============ benchmark_utcfromtimestamp ============
datetime_utcfromtimestamp 0.08515822299523279
windup_utcfromtimestamp 0.04857380600878969
windup is 1.8 times faster
```


CHANGELOG.md
* 0.1.1 (in beta) 2022.07.16
    * Complete comment


test/test_windup.py
import unittest
from datetime import datetime, timedelta, tzinfo

import windup

NO_DST = timedelta(0)


class Test(unittest.TestCase):

    def test_utcnow(self):
        dt_now = datetime.utcnow()
        wd_now = windup.utcnow()

        self.assertIsInstance(wd_now, datetime)

        self.assertEqual(wd_now.year, dt_now.year)
        self.assertEqual(wd_now.month, dt_now.month)
        self.assertEqual(wd_now.day, dt_now.day)

        self.assertEqual(wd_now.hour, dt_now.hour)
        self.assertEqual(wd_now.minute, dt_now.minute)
        self.assertEqual(wd_now.second, dt_now.second)

        self.assertEqual(wd_now.utcoffset(), timedelta(0))
        self.assertEqual(wd_now.dst(), NO_DST)

    def test_now(self):
        dt_now = datetime.now()
        wd_now = windup.now()

        self.assertIsInstance(wd_now, datetime)

        self.assertEqual(wd_now.year, dt_now.year)
        self.assertEqual(wd_now.month, dt_now.month)
        self.assertEqual(wd_now.day, dt_now.day)

        self.assertEqual(wd_now.hour, dt_now.hour)
        self.assertEqual(wd_now.minute, dt_now.minute)
        self.assertEqual(wd_now.second, dt_now.second)

    def test_from_and_to_string(self):
        rfc3339 = '2016-07-15T12:33:20.123000+01:30'
        dt = windup.from_string(rfc3339)

        self.assertIsInstance(dt, datetime)

        self.assertEqual(dt.year, 2016)
        self.assertEqual(dt.month, 7)
        self.assertEqual(dt.day, 15)

        self.assertEqual(dt.hour, 12)
        self.assertEqual(dt.minute, 33)
        self.assertEqual(dt.second, 20)
        self.assertEqual(dt.microsecond, 123000)

        self.assertEqual(dt.utcoffset(), timedelta(hours=1, minutes=30))
        self.assertEqual(dt.dst(), NO_DST)

        self.assertEqual(windup.to_string(dt), rfc3339)

        rfc3339 = '2016-07-18T12:58:26.485897-02:00'
        dt = windup.from_string(rfc3339)
        self.assertEqual(windup.to_string(dt), rfc3339)

    def test_fromtimestamp(self):
        DAY = 86400
        HOUR = 3600
        TZ_CEST = windup.TZFixedOffset(60 * 2)

        for t in range(0, DAY - (2 * HOUR), HOUR):
            dt = datetime.fromtimestamp(t)
            wdt = windup.from_timestamp(t)

            self.assertIsInstance(wdt, datetime)
            self.assertEqual(wdt.year, dt.year)
            self.assertEqual(wdt.month, dt.month)
            self.assertEqual(wdt.day, dt.day)

            self.assertEqual(wdt.hour, dt.hour)
            self.assertEqual(wdt.minute, dt.minute)
            self.assertEqual(wdt.second, dt.second)
            self.assertEqual(wdt.microsecond, dt.microsecond)

            self.assertEqual(wdt.utcoffset(), timedelta(0))
            self.assertEqual(wdt.dst(), NO_DST)

        for t in range(0, DAY, HOUR):
            dt = datetime.fromtimestamp(t, TZ_CEST)
            wdt = windup.from_timestamp(t, TZ_CEST)

            self.assertIsInstance(wdt, datetime)
            self.assertEqual(wdt.year, dt.year)
            self.assertEqual(wdt.month, dt.month)
            self.assertEqual(wdt.day, dt.day)

            self.assertEqual(wdt.hour, dt.hour)
            self.assertEqual(wdt.minute, dt.minute)
            self.assertEqual(wdt.second, dt.second)
            self.assertEqual(wdt.microsecond, dt.microsecond)

            self.assertEqual(wdt.utcoffset(), timedelta(hours=2))
            self.assertEqual(wdt.dst(), NO_DST)

        for t in range(0, DAY * -1, HOUR * -1):
            dt = datetime.fromtimestamp(t, TZ_CEST)
            wdt = windup.from_timestamp(t, TZ_CEST)

            self.assertIsInstance(wdt, datetime)
            self.assertEqual(wdt.year, dt.year)
            self.assertEqual(wdt.month, dt.month)
            self.assertEqual(wdt.day, dt.day)

            self.assertEqual(wdt.hour, dt.hour)
            self.assertEqual(wdt.minute, dt.minute)
            self.assertEqual(wdt.second, dt.second)
            self.assertEqual(wdt.microsecond, dt.microsecond)

            self.assertEqual(wdt.utcoffset(), timedelta(hours=2))
            self.assertEqual(wdt.dst(), NO_DST)

    def test_utcfromtimestamp(self):
        DAY = 86400
        HOUR = 3600

        for t in range(0, DAY, HOUR):
            dt = datetime.utcfromtimestamp(t)
            wdt = windup.from_utctimestamp(t)

            self.assertIsInstance(wdt, datetime)
            self.assertEqual(wdt.year, dt.year)
            self.assertEqual(wdt.month, dt.month)
            self.assertEqual(wdt.day, dt.day)
            self.assertEqual(wdt.hour, dt.hour)
            self.assertEqual(wdt.minute, dt.minute)
            self.assertEqual(wdt.second, dt.second)
            self.assertEqual(wdt.microsecond, dt.microsecond)

            self.assertEqual(wdt.utcoffset(), timedelta(0))
            self.assertEqual(wdt.dst(), NO_DST)

        for t in range(0, DAY * -1, HOUR * -1):
            dt = datetime.utcfromtimestamp(t)
            wdt = windup.from_utctimestamp(t)

            self.assertIsInstance(wdt, datetime)
            self.assertEqual(wdt.year, dt.year)
            self.assertEqual(wdt.month, dt.month)
            self.assertEqual(wdt.day, dt.day)
            self.assertEqual(wdt.hour, dt.hour)
            self.assertEqual(wdt.minute, dt.minute)
            self.assertEqual(wdt.second, dt.second)
            self.assertEqual(wdt.microsecond, dt.microsecond)

            self.assertEqual(wdt.utcoffset(), timedelta(0))
            self.assertEqual(wdt.dst(), NO_DST)

    def test_broken_from_string(self):
        invalid = [
            '2016-07-15 12:33:20.123000+01:30',
            '2016-13-15T12:33:20.123000+01:30',
            '20161315T12:33:20.123000+01:30',
            'Hello World',
            '2016-07-15 12:33:20.123000+01:302016-07-15 12:33:20.123000+01:30',
            '2016-07-15T12:33:20.1Z0',
            '2016-07-15T12:33:20.1 +01:30f',
        ]

        for r in invalid:
            with self.assertRaises(ValueError):
                windup.from_string(r)

    def test_ok_from_string(self):
        rfc3339s = [
            '2016-07-15 T 12:33:20.123000 +01:30',
            '2016-07-15 T 12:33:20.123000 +01:30',
            '2016-07-15T12:33:20.123 +01:30',
            '2016-07-15T12:33:20 +01:30',
            '2016-07-15T12:33:20 Z',
            '2016-07-15T12:33:20',
            '2016-07-15t12:33:20',
            '2016-07-15T12:33:20.1 +01:30',
        ]

        for r in rfc3339s:
            self.assertIsInstance(
                windup.from_string(r),
                datetime
            )

    def test_tzone(self):
        rfc3339 = '2016-07-15T12:33:20.123000+01:30'
        dt = windup.from_string(rfc3339)
        offset = dt.tzinfo.utcoffset()
        dst = dt.tzinfo.dst()

        self.assertIsInstance(offset, timedelta)
        self.assertEqual(offset.total_seconds() / 60, 90)
        self.assertEqual(dst, NO_DST)

        rfc3339 = '2016-07-15T12:33:20.123000Z'
        dt = windup.from_string(rfc3339)
        offset = dt.tzinfo.utcoffset()
        dst = dt.tzinfo.dst()

        self.assertIsInstance(offset, timedelta)
        self.assertEqual(offset.total_seconds(), 0)
        self.assertEqual(dst, NO_DST)

        rfc3339 = '2016-07-15T12:33:20.123000-02:00'
        dt = windup.from_string(rfc3339)
        offset = dt.tzinfo.utcoffset()
        dst = dt.tzinfo.dst()

        self.assertIsInstance(offset, timedelta)
        self.assertEqual(offset.total_seconds() / 60, -120)
        self.assertEqual(dst, NO_DST)

    def test_precision(self):
        t = 1469897308.549871
        dt = datetime.fromtimestamp(t)
        wdt = windup.from_timestamp(t)
        self.assertEqual(wdt.microsecond, dt.microsecond)

    def test_raise_on_not_TZFixedOffset(self):
        class TZInvalid(tzinfo):
            def utcoffset(self, dt=None):
                return timedelta(seconds=0)

            def dst(self, dt=None):
                return timedelta(seconds=0)

        dt = datetime.now(TZInvalid())

        with self.assertRaises(ValueError):
            windup.to_string(dt)

    def test_variable_fraction(self):
        rfc3339 = '2016-07-15T12:33:20.1'
        d1 = windup.from_string(rfc3339 + ('0' * 5) + 'Z')

        for x in range(0, 6):
            d2 = windup.from_string(rfc3339 + ('0' * x) + 'Z')
            self.assertEqual(d1, d2)

        self.assertEqual(
            windup.from_string('2016-07-15T12:33:20.123Z'),
            windup.from_string('2016-07-15T12:33:20.123000Z'),
        )

        self.assertEqual(
            windup.from_string('2016-07-15T12:33:20.0Z'),
            windup.from_string('2016-07-15T12:33:20Z'),
        )


if __name__ == '__main__':
    unittest.main()


scripts/bench_windup.py
from time import time
from datetime import datetime

import windup


RFC3339_DATE = '2016-07-18'
RFC3339_TIME = '12:58:26.485897+02:00'
RFC3339_DATE_TIME = RFC3339_DATE + 'T' + RFC3339_TIME
RFC3339_DATE_TIME_DTLIB = RFC3339_DATE_TIME[:-6]
DATE_TIME_FORMAT = '%Y-%m-%dT%H:%M:%S.%f'
DATETIME_OBJ = datetime.strptime(RFC3339_DATE_TIME_DTLIB, DATE_TIME_FORMAT)
TIME = time()
WINDUP_OPT = windup.fmt.date | windup.fmt.time | windup.fmt.usec


def benchmark_parse():
    def datetime_strptime():
        datetime.strptime(RFC3339_DATE_TIME_DTLIB, DATE_TIME_FORMAT)

    def windup_parse():
        windup.from_string(RFC3339_DATE_TIME)

    return datetime_strptime, windup_parse


def benchmark_format():
    def datetime_strftime():
        DATETIME_OBJ.strftime(DATE_TIME_FORMAT)

    def windup_format():
        windup.to_string(DATETIME_OBJ, option=WINDUP_OPT)

    return datetime_strftime, windup_format


def benchmark_utcnow():
    def datetime_utcnow():
        datetime.utcnow()

    def windup_utcnow():
        windup.utcnow()

    return datetime_utcnow, windup_utcnow


def benchmark_now():
    def datetime_now():
        datetime.now()

    def windup_now():
        windup.now()

    return datetime_now, windup_now


def benchmark_utcnow_to_string():
    def datetime_utcnow_to_string():
        datetime.utcnow().strftime(DATE_TIME_FORMAT)

    def windup_utcnow_to_string():
        windup.utcnow_to_string(option=WINDUP_OPT)

    return datetime_utcnow_to_string, windup_utcnow_to_string


def benchmark_now_to_string():
    def datetime_now_to_string():
        datetime.now().strftime(DATE_TIME_FORMAT)

    def windup_now_to_string():
        windup.now_to_string(option=WINDUP_OPT)

    return datetime_now_to_string, windup_now_to_string


def benchmark_fromtimestamp():
    def datetime_fromtimestamp():
        datetime.fromtimestamp(TIME)

    def windup_fromtimestamp():
        windup.from_timestamp(TIME)

    return datetime_fromtimestamp, windup_fromtimestamp


def benchmark_utcfromtimestamp():
    def datetime_utcfromtimestamp():
        datetime.utcfromtimestamp(TIME)

    def windup_utcfromtimestamp():
        windup.from_utctimestamp(TIME)

    return datetime_utcfromtimestamp, windup_utcfromtimestamp


if __name__ == '__main__':
    from timeit import repeat

    benchmarks = [
        benchmark_parse,
        benchmark_format,

        benchmark_utcnow,
        benchmark_now,

        benchmark_utcnow_to_string,
        benchmark_now_to_string,

        benchmark_fromtimestamp,
        benchmark_utcfromtimestamp,
    ]

    print('Executing benchmarks ...')

    for k in benchmarks:
        print('\n============ %s ============' % k.__name__)
        mins = []

        for func in k():
            times = repeat(stmt='func()',
                           setup='from __main__ import func',
                           number=100000)
            t = min(times)
            mins.append(t)
            print(func.__name__, t)

        win = False
        if mins[0] > mins[1]:
            win = True

        mins = sorted(mins)
        diff = mins[1] / mins[0]

        if win:
            print('windup is %.01f times faster' % diff)
        else:
            print('windup is %.01f times slower' % diff)


windup/format.py
from typing import Union
from collections import namedtuple
from datetime import datetime, date, time as dtime

from windup.rfc3339 import (
    str_usec_to_msec,
    utcnow_to_string as _u_utcnow_to_string,
    now_to_string as _u_now_to_string,
    to_string as _u_to_string
)

# date-time string format separator
sep = namedtuple('sep', ['T', 'space', 'underscore'])('T', ' ', '_')

# date-time string format option
fmt = namedtuple(
    'fmt',
    ['date', 'time', 'msec', 'usec', 'tz', 'utc_z']
)(1, 1 << 1, 1 << 2, 1 << 3, 1 << 4, 1 << 5)

default_fmt = fmt.date | fmt.time | fmt.usec | fmt.tz


def _to_string_fmt(dt_string: str, option: int, sep: str) -> str:
    """Format date-time string."""
    _dts, _no_date, _no_time = '', 0, 0

    if option & fmt.date != 0:
        _dts = dt_string[:10]
    else:
        _no_date = 1

    if option & fmt.time != 0:
        if _no_date:
            _dts += dt_string[11:19]
        else:
            _dts += sep + dt_string[11:19]
    else:
        _no_time = 1

    if option & fmt.usec != 0:
        if _no_time:
            _dts += dt_string[20:26]
        else:
            _dts += dt_string[19:26]
    elif option & fmt.msec != 0:
        if _no_time:
            _dts += str_usec_to_msec(dt_string[20:26])
        else:
            _dts += '.' + str_usec_to_msec(dt_string[20:26])

    if option & fmt.utc_z != 0:
        if dt_string[-6:] in ('+00:00', '-00:00'):
            _dts += 'Z'
        else:
            _dts += dt_string[-6:]
    elif option & fmt.tz != 0:
        _dts += dt_string[-6:]

    return _dts


def _dt_to_string(dt: datetime, option: int, sep: str) -> str:
    """Serialize datetime object to date-time string."""
    return _to_string_fmt(_u_to_string(dt), option=option, sep=sep)


def utcnow_to_string(option: int = default_fmt, sep: str = sep.T) -> str:
    """
    Serialize utcnow datetime object to date-time string.

    Args:
        option: date-time string format option.
                    fmt.date  - format date string
                    fmt.time  - format time string
                    fmt.msec  - format millisecond string
                    fmt.usec  - format microseconds string
                    fmt.tz    - format timezone string
                    fmt.utc_z - use 'Z' instead of '+00:00'
                                if the timezone is UTC.

        sep:    date-time string separator between date and time.
                    sep.T          - 'T'
                    sep.space      - ' '
                    sep.underscore - '_'

    Returns:
        date-time string

    Examples:
        >>> import windup
        >>> windup.utcnow_to_string()
        '2022-07-12T11:18:55.547682+00:00'
        >>> windup.utcnow_to_string(option=windup.fmt.date | windup.fmt.time | windup.fmt.msec)
        '2022-07-13T08:41:15.397'
        >>> windup.utcnow_to_string(option=windup.fmt.date | windup.fmt.time, sep=windup.sep.space)
        '2022-07-13 08:42:32'

    """
    return _to_string_fmt(_u_utcnow_to_string(), option=option, sep=sep)


def now_to_string(option: int = default_fmt, sep: str = sep.T) -> str:
    """
    Serialize now datetime object to date-time string.

    Args:
        option: date-time string format option.
                    fmt.date  - format date string
                    fmt.time  - format time string
                    fmt.msec  - format millisecond string
                    fmt.usec  - format microseconds string
                    fmt.tz    - format timezone string
                    fmt.utc_z - use 'Z' instead of '+00:00'
                                if the timezone is UTC.

        sep:    date-time string separator between date and time.
                    sep.T          - 'T'
                    sep.space      - ' '
                    sep.underscore - '_'

    Returns:
        date-time string

    Examples:
        >>> import windup
        >>> windup.now_to_string()
        '2022-07-12T11:18:55.547000+00:00'
        >>> windup.now_to_string(option=windup.fmt.date | windup.fmt.time | windup.fmt.msec)
        '2022-07-13T16:46:25.120'
        >>> windup.now_to_string(option=windup.fmt.date | windup.fmt.time, sep=windup.sep.space)
        '2022-07-13 16:47:03'

    """
    return _to_string_fmt(_u_now_to_string(), option=option, sep=sep)


def to_string(dt: Union[datetime, date, dtime],
              option: int = default_fmt,
              sep: str = sep.T) -> str:
    """
    Serialize datetime object to date-time string.

    Args:
        dt:     datetime object.
        option: date-time string format option.
                    fmt.date  - format date string
                    fmt.time  - format time string
                    fmt.msec  - format millisecond string
                    fmt.usec  - format microseconds string
                    fmt.tz    - format timezone string
                    fmt.utc_z - use 'Z' instead of '+00:00'
                                if the timezone is UTC.

        sep:    date-time string separator between date and time.
                    sep.T          - 'T'
                    sep.space      - ' '
                    sep.underscore - '_'

    Returns:
        date-time string

    Raises:
        TypeError

    Examples:
        >>> import windup
        >>> windup.to_string(windup.utcnow())
        '2022-07-12T11:18:55.547000+00:00'
        >>> windup.to_string(windup.utcnow(), option=windup.fmt.date | windup.fmt.time | windup.fmt.usec)
        '2022-07-13T08:48:43.524185'
        >>> windup.to_string(windup.utcnow(), option=windup.fmt.date | windup.fmt.time | windup.fmt.utc_z, sep=windup.sep.space)
        '2022-07-13 08:49:21Z'

    """
    _dt = None
    if isinstance(dt, datetime):
        _dt = dt
    elif isinstance(dt, date):
        _dt = datetime(dt.year, dt.month, dt.day)
        option = fmt.date
    elif isinstance(dt, dtime):
        _dt = datetime(
            1970, 1, 1,
            dt.hour, dt.minute, dt.second, dt.microsecond,
            tzinfo=dt.tzinfo
        )
        if option & fmt.date != 0:
            option ^= fmt.date
    else:
        raise TypeError('Expected a datetime/date/time object.')

    return _dt_to_string(_dt, option=option, sep=sep)
